price_feed/lib.rs

```diff
@@ -0,0 +1,197 @@
+    pub enum PalletAssetArithmeticErr {
+        /// Underflow.
+        Underflow,
+        /// Overflow.
+        Overflow,
+        /// Division by zero.
+        DivisionByZero,
+        // unknown error
+        Unknown,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum PalletAssetTokenErr {
+        /// Funds are unavailable.
+        NoFunds,
+        /// Account that must exist would die.
+        WouldDie,
+        /// Account cannot exist with the funds that would be given.
+        BelowMinimum,
+        /// Account cannot be created.
+        CannotCreate,
+        /// The asset in question is unknown.
+        UnknownAsset,
+        /// Funds exist but are frozen.
+        Frozen,
+        /// Operation is not supported by the asset.
+        Unsupported,
+        // unknown error
+        Unknown,
+    }
+
+    impl ink::env::chain_extension::FromStatusCode for ChainExtensionError {
+        fn from_status_code(status_code: u32) -> Result<(), Self> {
+            match status_code {
+                0 => Ok(()),
+                _ => panic!("encountered unknown status code"),
+            }
+        }
+    }
+
+    mod tests {
+        /// Imports all the definitions from the outer scope so we can use them here.
+        use super::*;
+
+        /// We test a simple use case of our contract.
+        #[ink::test]
+        fn it_works() {
+            let mut price_feed = PriceFeed::new();
+            assert_eq!(
+                price_feed.get_coin_info([0u8; 32], [0u8; 32]),
+                Ok(CoinInfo {
+                    symbol: vec![],
+                    name: vec![],
+                    blockchain: vec![],
+                    supply: 0,
+                    last_update_timestamp: 0,
+                    price: 0
+                })
+            );
+        }
+    }
+
+    /// This is how you'd write end-to-end (E2E) or integration tests for ink! contracts.
+    ///
+    /// When running these you need to make sure that you:
+    /// - Compile the tests with the `e2e-tests` feature flag enabled (`--features e2e-tests`)
+    /// - Are running a Substrate node which contains `pallet-contracts` in the background
+    #[cfg(all(test, feature = "e2e-tests"))]
+    mod e2e_tests {
+        /// Imports all the definitions from the outer scope so we can use them here.
+        use super::*;
+
+        /// A helper function used for calling contract messages.
+        use ink_e2e::build_message;
+
+        /// The End-to-End test `Result` type.
+        type E2EResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+        /// We test that we can upload and instantiate the contract using its default constructor.
+        #[ink_e2e::test]
+        async fn default_works(mut client: ink_e2e::Client<C, E>) -> E2EResult<()> {
+            // Given
+            let constructor = AaaRef::default();
```

Comment:
Where is `AaaRef` coming from? Was this taken from an example?

Reply:
Can we maybe use something else but 'aaa'? If I understand correctly this is a test instance of our contract?

Reply:
Sorry, this shouldn't have been included. The tests have not yet been implemented. There's another ticket to address this. Removed it from this PR.